File: python/deque/DequeDoubleLinkedList.py
class Node:
    """
    Class representing a single node in a doubly linked list.
    """
    def __init__(self, data):
        """
        Initialize a node with data and pointers to previous and next nodes.
        :param data: The value to be stored in the node.
        """
        self.data = data  # The value stored in the node
        self.prev = None  # Pointer to the previous node
        self.next = None  # Pointer to the next node


class DequeDoubleLinkedList:
    """
    Implementation of a Deque (double-ended queue) using a doubly linked list.
    """
    def __init__(self):
        """
        Initialize an empty deque.
        """
        self.head = None  # Pointer to the front of the deque
        self.tail = None  # Pointer to the rear of the deque
        self.size = 0     # Number of elements in the deque

    def append(self, data):
        """
        Add an element to the rear of the deque.
        :param data: The value to be added.
        """
        new_node = Node(data)  # Create a new node
        if not self.head:  # If the deque is empty
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail  # Link the new node to the current tail
            self.tail.next = new_node  # Link the current tail to the new node
            self.tail = new_node       # Update the tail pointer
        self.size += 1  # Increment the size

    def appendleft(self, data):
        """
        Add an element to the front of the deque.
        :param data: The value to be added.
        """
        new_node = Node(data)  # Create a new node
        if not self.head:  # If the deque is empty
            self.head = self.tail = new_node
        else:
            new_node.next = self.head  # Link the new node to the current head
            self.head.prev = new_node  # Link the current head to the new node
            self.head = new_node       # Update the head pointer
        self.size += 1  # Increment the size

    def pop(self):
        """
        Remove and return the element from the rear of the deque.
        :return: The value of the removed element.
        :raises IndexError: If the deque is empty.
        """
        if not self.tail:  # If the deque is empty
            raise IndexError("pop from empty deque")
        data = self.tail.data  # Store the value to be returned
        if self.head == self.tail:  # If there's only one element
            self.head = self.tail = None  # Reset head and tail to None
        else:
            self.tail = self.tail.prev  # Move the tail pointer back
            self.tail.next = None  # Unlink the last node
        self.size -= 1  # Decrement the size
        return data

    def popleft(self):
        """
        Remove and return the element from the front of the deque.
        :return: The value of the removed element.
        :raises IndexError: If the deque is empty.
        """
        if not self.head:  # If the deque is empty
            raise IndexError("pop from empty deque")
        data = self.head.data  # Store the value to be returned
        if self.head == self.tail:  # If there's only one element
            self.head = self.tail = None  # Reset head and tail to None
        else:
            self.head = self.head.next  # Move the head pointer forward
            self.head.prev = None  # Unlink the first node
        self.size -= 1  # Decrement the size
        return data

    def clear(self):
        """
        Remove all elements from the deque.
        """
        self.head = self.tail = None  # Reset head and tail pointers
        self.size = 0  # Reset size

    def __len__(self):
        """
        Return the number of elements in the deque.
        :return: The size of the deque.
        """
        return self.size

    def __iter__(self):
        """
        Iterate over the elements of the deque from front to rear.
        :yield: The data of each node in the deque.
        """
        current = self.head  # Start at the front of the deque
        while current:
            yield current.data  # Yield the value of the current node
            current = current.next  # Move to the next node

File: src/main/java/deque/DequeDoubleLinkedList.java
package deque;

/**
 * Class representing a doubly linked list node.
 */
class Node<T> {
    T data;       // Data stored in the node
    Node<T> next; // Pointer to the next node
    Node<T> prev; // Pointer to the previous node

    public Node(T data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

/**
 * Implementation of a Deque (Double-Ended Queue) using a Doubly Linked List.
 */
public class DequeDoubleLinkedList<T> {
    private Node<T> head; // Points to the front of the deque
    private Node<T> tail; // Points to the rear of the deque
    private int size;     // Tracks the number of elements in the deque

    /**
     * Constructor to initialize an empty deque.
     */
    public DequeDoubleLinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    /**
     * Add an element to the rear of the deque.
     * @param value The value to add.
     */
    public void append(T value) {
        Node<T> newNode = new Node<>(value);
        if (tail == null) { // If the deque is empty
            head = tail = newNode;
        } else {
            tail.next = newNode;
            newNode.prev = tail;
            tail = newNode;
        }
        size++;
    }

    /**
     * Add an element to the front of the deque.
     * @param value The value to add.
     */
    public void appendLeft(T value) {
        Node<T> newNode = new Node<>(value);
        if (head == null) { // If the deque is empty
            head = tail = newNode;
        } else {
            newNode.next = head;
            head.prev = newNode;
            head = newNode;
        }
        size++;
    }

    /**
     * Remove and return the element from the rear of the deque.
     * @return The value of the removed element.
     * @throws IllegalStateException if the deque is empty.
     */
    public T pop() {
        if (tail == null) {
            throw new IllegalStateException("Deque is empty!");
        }
        T value = tail.data;
        if (head == tail) { // If there's only one element
            head = tail = null;
        } else {
            tail = tail.prev;
            tail.next = null;
        }
        size--;
        return value;
    }

    /**
     * Remove and return the element from the front of the deque.
     * @return The value of the removed element.
     * @throws IllegalStateException if the deque is empty.
     */
    public T popLeft() {
        if (head == null) {
            throw new IllegalStateException("Deque is empty!");
        }
        T value = head.data;
        if (head == tail) { // If there's only one element
            head = tail = null;
        } else {
            head = head.next;
            head.prev = null;
        }
        size--;
        return value;
    }

    /**
     * Check if the deque is empty.
     * @return True if the deque is empty, false otherwise.
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Get the current size of the deque.
     * @return The number of elements in the deque.
     */
    public int getSize() {
        return size;
    }

    /**
     * Print all elements in the deque from front to rear.
     */
    public void printDeque() {
        Node<T> current = head;
        while (current != null) {
            System.out.print(current.data + " <-> ");
            current = current.next;
        }
        System.out.println("null");
    }

    /**
     * Main method to demonstrate the Deque operations.
     */
}
